app/app.rb

```diff
@@ -61,15 +63,32 @@ class App < Base
     ))
   end
 
+  # Marks an invoice as processed
+  # Processing has to be performed externally and can be anything,
+  # e.g. email the invoice to the customer.
+  #
+  # number - The invoice number.
+  #
+  # Returns a JSON representation of the invoice
+  post '/invoices/:number/process' do
+    invoice = Invoice.where(number: params[:number]).first
+
+    halt 404 unless invoice
+    halt 400 if !invoice.pdf_generated? || invoice.processed?
```

Comment:
We probably shouldn't add the pdf generated constraint here if the unprocessed scope doesn't filter invoices on pdf generated. I would be inclined to remove this check and make processing a processed invoice a no-op. The client is then responsible for asking invoices that are pdf-generated + are unprocessed, making this more general.

We then probably also don't need the `ProcessingError`, as the client would be responsible for fetching only invoices that are processed as well. Keeps it simple.

Thoughts?

Reply:
Sure, I can follow your train of thought there. I envisioned more of a fixed progression of states, but leaving it open to the client works just as well and might allow for more use cases then those I had in mind.